### PR TITLE
Default network implementation

### PR DESCRIPTION
**Pull request by milosgajdos**
This PR is a WIP version of default network implementation.

### CODE REVIEW

File: network/default.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+	if err != nil {
+		return err
+	}
+
+	// advertise routes
+	go n.advertise(client, advertChan)
+	// process routes
+	go n.process(client)
+
+	// set connected to true
```
Comment (asim) [inline network/default.go:337]: Start the server

File: network/default.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+	n.connected = true
+
+	return nil
+}
+
+func (n *network) close() error {
+	// stop the router
```
Comment (asim) [inline network/default.go:349]: Stop the server

File: network/default.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	)
+
+	// srv is network server
+	srv := server.NewServer(
+		server.Transport(tunTransport),
+	)
+
+	// client is network client
+	client := client.NewClient(
+		client.Transport(tunTransport),
```
Comment (asim) [inline network/default.go:73]: Set the router selector

```
import "github.com/micro/go-micro/client/selector/router"

client.Selector(
    router.NewSelector(
        router.WithRouter(r),
    ),
)
```

File: network/network.go [hunk 4]
```diff
@@ -1,2 +1,37 @@
...
+	DefaultAddress = ":0"
+	// ResolveTime ddefines the time we periodically resolve network nodes
+	ResolveTime = 1 * time.Minute
+)
+
+// Network is micro network
+type Network interface {
+	// Name of the network
+	Name() string
+	// Connect starts the resolver and tunnel server
```
Comment (asim) [inline network/network.go:26]: Add `Address() string` method to the network interface to expose the `tunnel.Address()`

File: network/network.go [hunk 5]
```diff
@@ -1,2 +1,37 @@
...
+import (
+	"time"
+
+	"github.com/micro/go-micro/client"
+	"github.com/micro/go-micro/server"
+)
+
+var (
+	// DefaultName is default network name
+	DefaultName = "go.micro.network"
```
Comment (asim) [inline network/network.go]: Set network name to `go.micro`

File: network/default.go [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+	)
+
+	// create tunnel client with tunnel transport
+	tunTransport := transport.NewTransport(
+		tr.WithTunnel(options.Tunnel),
+	)
+
+	// srv is network server
+	srv := server.NewServer(
+		server.Transport(tunTransport),
```
Comment (asim) [inline network/default.go:68]: Set server name to the same as the network name

File: network/default.go [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+// Name returns network name
+func (n *network) Name() string {
+	return n.options.Name
+}
+
+// Address returns network bind address
+func (n *network) Address() string {
+	return n.options.Address
```
Comment (asim) [inline network/default.go]: return n.tunnel.Address()

File: network/default.go [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+		return err
+	}
+
+	// initialize the tunnel to resolved nodes
+	n.Tunnel.Init(
+		tunnel.Nodes(nodes...),
+	)
+
+	// dial into ControlChannel to send route adverts
+	client, err := n.Tunnel.Dial(ControlChannel)
```
Comment (asim) [inline network/default.go:299]: Should we also be listening on the control channel?

File: network/default.go [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+	"github.com/micro/go-micro/server"
+	"github.com/micro/go-micro/transport"
+	"github.com/micro/go-micro/tunnel"
+	tr "github.com/micro/go-micro/tunnel/transport"
+	"github.com/micro/go-micro/util/log"
+)
+
+var (
+	// ControlChannel is the name of the tunnel channel for passing contron message
+	ControlChannel = "control-msg"
```
Comment (asim) [inline network/default.go]: Should we name this to `router` or something similar?

File: network/default.go (hunk 8 above)
Comment (milosgajdos) [reply network/default.go:299]: Yes we should! 😜

File: network/default.go (hunk 1 above)
Comment (milosgajdos) [reply network/default.go:337]: Should we consider the router to be started or should we start the router here, too?

File: network/default.go (hunk 2 above)
Comment (milosgajdos) [reply network/default.go:349]: Ditto. Should we stop the router here as well?

File: network/default.go (hunk 9 above)
Comment (milosgajdos) [reply network/default.go]: That depends: are we going to be sending **ONLY router** adverts on this or are we going to re-use this particular `tunnel` channel for other "control" messages?

File: network/default.go (hunk 1 above)
Comment (asim) [reply network/default.go:337]: We need to start the router

File: network/default.go (hunk 2 above)
Comment (asim) [reply network/default.go:349]: Yea

File: network/default.go (hunk 9 above)
Comment (asim) [reply network/default.go]: Let's just call it control and reuse it for now. Probably keeps things simple.

File: network/default.go (hunk 1 above)
Comment (milosgajdos) [reply network/default.go:337]: I need to make the router start idempotent first 😅

Comment (asim) [reply network/default.go:337]: You have the status that will tell you if its running right?

Comment (milosgajdos) [reply network/default.go:337]: Yeah, but that code needs to be written. On it now.

Comment (asim) [reply network/default.go:337]: Pushed that change

File: network/default.go (hunk 8 above)
Comment (milosgajdos) [reply network/default.go:299]: Actually...should we? 🧐

File: network/default.go [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+		case advert := <-advertChan:
+			// create a proto advert
+			var events []*pb.Event
+			for _, event := range advert.Events {
+				route := &pb.Route{
+					Service: event.Route.Service,
+					Address: event.Route.Address,
+					Gateway: event.Route.Gateway,
+					Network: event.Route.Network,
+					Link:    event.Route.Link,
```
Comment (asim) [inline network/default.go]: Should we set the link here to "network" or on receiving?

Comment (milosgajdos) [reply network/default.go]: Not sure. On send we set it to "network" saying, hey this is a "remote" route. On receive you should just process the routes as you received them, no?

Comment (asim) [reply network/default.go]: Network field should not be modified. Link field should be set to "Network". This may be why it's not working.

Comment (milosgajdos) [reply network/default.go]: Yeah I got Network and Link mixed up -- my brain short-circuited the string "network" with `Route.Network` lmao. Nice one!

**Pull request merged**